### DISCUSSION TITLE
sending temperatur sensor data with MQTT when temperature changes

### DISCUSSION THREAD

**Discussion by roesl** [arendst/Tasmota#17198] [Ideas]
Hello,

is it possible, to send temp data ( DS18B20) with mqtt command when temp changes similar to function of Shelly Uni. There is a menu, where you can configure a mqtt command which is sent to a specific device, when temp changes.

Is it possible to do this with a Berry script  on a ESP32 since there is no function in tasmota ?

**Comment by sfromis** [comment]
Yes, in Berry it does not take much code to have a program calling `tasmota.read_sensors` at short intervals, detect differences, and do whatever, like a publish.

You can also do something like that [with classic rules](https://tasmota.github.io/docs/Rules/#transmit-sensor-value-only-when-a-delta-is-reached), but when you have ESP32, I'd much recommend doing it in Berry.

**Comment by roesl** [comment]
Hi, It would be better, when data sent when changing. Also i would prefere to make config similar to shelly uni, so that config can made from normal users. --Diese Nachricht wurde von meinem Android Mobiltelefon mit 1&1 Mail gesendet.Am 27.11.22, 11:32 schrieb sfromis ***@***.***>:

   
  Yes, in Berry it does not take much code to have a program calling tasmota.read_sensors at short intervals, detect differences, and do whatever, like a publish. 
  You can also do something like that with classic rules, but when you have ESP32, I'd much recommend doing it in Berry.

**Comment by roesl** [comment]
Hi, 

I tried with Rules  

DS18B20#Temperaturewhenever the temperature of sensor DS18B20 changes

Rule1
on DS18B20#temperature do backlog var1 %value%; sub1 5; event sendtemp endon
on event#sendtemp do publish shellies/shellytrv-8CF681148755/temp %var1% endon

 

but the temp is sent every second but not only on changes 

 

Whats wrong ?

**Reply by barbudor** [reply]
You just wrote a rule to send the value of temp-5 every time temp is measured

Read again the example provided by sfromis

**Reply by sfromis** [reply]
You did not read the example I posted a link to, but instead broke something from the next example (already a bit bad in a different way), which is about changing the value, not detecting differences.

Anyway, I was playing with a bit of Berry code to publish in case of sensor value delta over a selected value. `SensorDelta.be`:
```
class SensorDelta
  var deltatuples,prevvalues,name
  def init(intervalseconds, deltatuples)
    import string,json,mqtt
    self.deltatuples = deltatuples
    var times = 0
    self.name = classname(self)
    var topic = string.replace(string.replace(tasmota.cmd('FullTopic')['FullTopic'], '%topic%', tasmota.cmd('Topic')['Topic']), '%prefix%', tasmota.cmd('Prefix')['Prefix3']) + 'SENSOR'
    self.prevvalues = {}
    tasmota.add_rule('Time', / value,trigger,jsonmap -> self.onecheck(jsonmap, true))
    tasmota.remove_cron(self.name)
    tasmota.add_cron('*/' + str(intervalseconds) + ' * * * * *',
      def()
        times -= 1
        if times == 0
          tasmota.remove_cron(self.name)
        end
        var sensorsjson = tasmota.read_sensors()
        if self.onecheck(json.load(sensorsjson), false)
          mqtt.publish(topic, sensorsjson[0..size(sensorsjson)-2] + ',"' + self.name + '":true}')
        end
      end,
      self.name)
  end
  def onecheck(sensorsmap, update)
    import math
    if sensorsmap.find(self.name, false) return false end
    var exceeded = false
    for sensorname : sensorsmap.keys()
      var entry = sensorsmap[sensorname]
      if classname(entry) == 'map'
        for key : entry.keys()
          var value = entry[key]
          if value == nil break end
          var fullkey = sensorname + '.' + key
          var maxdelta = self.deltatuples.find(fullkey, -1)
          if maxdelta == -1 maxdelta = self.deltatuples.find(key, -1) end
          if maxdelta == -1 continue end
          var prevvalue = self.prevvalues.find(fullkey, nil)
          if prevvalue == nil
            prevvalue = value
            update = true
          end
          var delta = value - prevvalue
          exceeded = exceeded || math.abs(delta) >= maxdelta
          if exceeded || update
            #if exceeded print(fullkey, 'delta', delta) end
            self.prevvalues.setitem(fullkey, value)
          end
        end
      end
    end
    return exceeded
  end
end
var sensordelta = SensorDelta(2, {'Temperature':0.2, 'BME280.Humidity':5})
```
This is generalized to be able to handle several sensors, with separate delta thresholds for each. The last line is an example of how it works, at 2 second intervals it will detect differences of all Temperature values (if having multiple sensors), and Humidity from the BME280 sensor, and publish if either exceeds the specified max delta. Obviously adapt to your case.

I do not see it as a good idea to simply check for the value being different, as sensors commonly may have jitter in the value produced, hence my use of a delta threshold. If you want to have all the jitter result in publish, setting a max delta of 0 works for that.

You can put it in the file system as `SensorDelta.be` and from `autoexec.be` do a `load('SensorDelta.be')`. Or just put the code directly in `autoexec.be`

**Reply by roesl** [reply]
thanks, I will try it with Berry.

But how I can set a offset value wihin this rule 👍 
Rule2
  ON DS18B20#temperature>%var1% DO Backlog var1 %value%; publish shellies/shellytrv-8CF681148755/temp %value%; var2 %value%; add1 0.3; sub2 0.3 ENDON
  ON DS18B20#temperature<%var2% DO Backlog var2 %value%; publish shellies/shellytrv-8CF681148755/temp %value%; var1 %value%; add1 0.3; sub2 0.3 ENDON

**Reply by sfromis** [reply]
You were not asking for an offset, but about detecting temperature changes. The thing with an offset is completely different, and you do neither need rules nor Berry for this, as you have the command `TempOffset` directly for this purpose, no hassle

**Comment by lalo-uy** [comment]
He use the name offset, put he refers to Temp Delta.

**Reply by roesl** [reply]
I need an offset of +20 for outside temperature to get positive values for writing  the value  to the thermostat "tempTargetSet" of Tasmota and offset of -10 for forward temp of floor heating, because the Shelly TRV only can set up to 31 degrees C for set temp.
then I can calculate the necessary forward temp for heating depending on  outside temp  and  write the value to the settemp of Shelly TRV
Also I need Temp Delta of +- 0.5 to reduce traffic.
